XBlockAside improvement. Difficulty selectbox in Studio (based on new XBlockAside functionality)

StraNNiKK opened the pull request:
XBlockAside improvement include:
- adaptation asides to be imported from the XML
- updating SplitMongo to handle XBlockAsides (CRUD operations). It means that now asides could be stored in MongoDB in this way:
```python
asides = item.runtime.get_asides(item)
asides[n].some_field = 'new_value'
modulestore().store.update_item(item, user_id, asides=[asides[n]])
```
- updating Studio to route some HTTP calls directly to XBlockAsides handlers
- updating xblock/core.js to properly init XBlockAsides connected JS files

Intentionally not yet implemented, coming in a subsequent PR:
1. The existed values for the Difficulty selectbox are the same for all courses and all problems and defined like as constant values in code. As I understand in future they will be taken from some external service - tagging server or from somewhere else.
2. XblockAsides could be imported from the course XML but could not be exported to XML.

edx-webhook commented:
Thanks for the pull request, @StraNNiKK! I've created [OSPR-1149](https://openedx.atlassian.net/browse/OSPR-1149) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

nedbat commented:
@StraNNiKK wow, thanks, this looks like a lot of work.  Have you talked to people here at edX about this work, and how it fits in to the plans for Asides?

cpennington commented:
@nedbat I reviewed @StraNNiKK's previous PR, and I know he's been chatting with @jbarciauskas about it on slack.

cpennington commented on cms/lib/xblock/tagging.py:
```diff
@@ -69,10 +75,16 @@ class StructuredTagsAside(XBlockAside):
 
     @XBlockAside.aside_for(STUDENT_VIEW)
-    def student_view_aside(self, block, context):
+    def student_view_aside(self, block, context):  # pylint: disable=W0612,W0613
```
Nit: please use the human read-able names for pylint messages.

cpennington commented on cms/lib/xblock/tagging.py:
```diff
@@ -86,7 +98,46 @@ def student_view_aside(self, block, context):
+        for path_part in request.path.split('/'):
+            if path_part.startswith(AsideUsageKeyV1.CANONICAL_NAMESPACE):
+                usage_key = UsageKey.from_string(path_part)
+                break
```
You shouldn't need to do this kind of parsing in this code.

cpennington commented on cms/lib/xblock/tagging.py:
```diff
@@ -86,7 +98,46 @@ def student_view_aside(self, block, context):
+
+        if usage_key:
+            descriptor = modulestore().get_item(usage_key.usage_key)
+            modulestore().update_item(descriptor, request._request.user.id, asides=[self])  # pylint: disable=protected-access
```
This is breaking an abstraction barrier. XBlocks shouldn't be calling Modulestore code. Better would be to have the code that is calling the handler on the aside call update_item and pass in the (possibly modified) aside.

StraNNiKK commented:
@cpennington I've updated the pull request according your comments

cpennington commented on cms/lib/xblock/test/test_tagging.py:
```diff
@@ -0,0 +1,151 @@
+        self.assertGreater(len(StructuredTagsAside.available_tags), 0,
+                           "StructuredTagsAside should contains at least one available tag")
+
+    def test_preivew_html(self):
```
Nit: typo. `s/preivew/preview/`

cpennington commented on cms/lib/xblock/test/test_tagging.py:
```diff
@@ -0,0 +1,151 @@
+        self.assertRegexpMatches(problem_html, r'<select\s[_=a-zA-Z0-9\"\']+\sname=[\"\']%s[\"\']' % self.aside_tag)
+        self.assertRegexpMatches(problem_html, r"data-block-type=[\"\']%s[\"\']" % self.aside_name)
+        self.assertRegexpMatches(problem_html, r"data-init=[\"\']StructuredTagsInit[\"\']")
+        self.assertRegexpMatches(problem_html, r"data-runtime-class=[\"\']PreviewRuntime[\"\']")
```
It would be better to test this by parsing the HTML with lxml, so that you can look for nodes with particular attributes, rather than using a regex (which is going to be vulnerable to ordering/spacing changes).

cpennington commented:
@StraNNiKK: This is excellent work! It's great to see so much forward progress on Asides.

One thing that would be really nice as you're adding functionality to the modulestores would be additional tests in https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/tests/test_semantics.py. These tests should use the modulestore interfaces (only the externally visible ones) to verify that split-mongo and draft-mongo behave the same.

StraNNiKK commented:
@cpennington hi. I've updated the pull requests. I've added additional semantics tests as you wrote.
All checks have failed - but it seems that it is jenkins issue.

cpennington commented on common/lib/xmodule/xmodule/modulestore/tests/test_semantics.py:
```diff
@@ -361,6 +412,32 @@ class TestSplitDirectOnlyCategorySemantics(DirectOnlyCategorySemantics):
+
+        self.assertBlockHasContent(block_usage_key, test_data.field_name, test_data.initial,
+                                   self.ASIDE_DATA_FIELD.field_name, self.ASIDE_DATA_FIELD.updated)
+
```
Can you move these methods up to `DirectOnlyCategorySemantics`, so that we verify them for both split-mongo and old-mongo? (Or are asides only applicable to split?)

StraNNiKK replied:
@cpennington Asides applicable only for split. As I know from edx JIRA you are not going to adapt asides for working with old mongo: https://openedx.atlassian.net/browse/PLAT-223

jbarciauskas replied:
@cpennington @StraNNiKK what's the next step on this? Is this waiting for you, Cale?

StraNNiKK replied:
@jbarciauskas I'm waiting for @cpennington

cpennington replied:
Oops. Sorry for leaving this hanging.

StraNNiKK commented:
jenkins run all

StraNNiKK commented:
jenkins run python

StraNNiKK commented:
jenkins run quality

StraNNiKK commented:
jenkins run js

cpennington commented on common/lib/xmodule/xmodule/modulestore/tests/test_semantics.py:
```diff
@@ -361,6 +418,32 @@ class TestSplitDirectOnlyCategorySemantics(DirectOnlyCategorySemantics):
+
+        with self.store.branch_setting(ModuleStoreEnum.Branch.draft_preferred):
+            block = self.store.get_item(block_usage_key)
+            aside = self._get_aside(block)
```
The intention of these tests is to work only using the public `Modulestore` interface. So, in this case, it would be better to assert that `block.get_asides()` returns the expected aside, rather than using the `_get_aside` method which isn't public.

StraNNiKK replied:
@cpennington but it is a helper method of the current test case. I mean it isn't the protected method of the block. In this case it uses as a helper function for the `block.runtime.get_asides(block)` construction. Of course, if you want I can use get asides without helper methods.

cpennington replied:
Oops. You're absolutely right, I was reading too quickly and missed that. Sorry. :+1:

cpennington commented:
@cptvitamin: This is Asides-related, and has user-facing changes.
